modules/3d-tiles/src/lib/tileset/base-tileset-traverser.js

```diff
@@ -0,0 +1,337 @@
+  async executeTraversal(root, frameState) {
+    // stack to store traversed tiles, only visible tiles should be added to stack
+    // visible: visible in the current view frustum
+    const stack = this._traversalStack;
+    stack.push(root);
+
+    while (stack.length > 0) {
+      // 1. pop tile
+      const tile = stack.pop();
+
+      // 2. check if tile needs to be refine, needs refine if a tile's LoD is not sufficient and tile has available children (available content)
+      let refines = false;
+      if (this.canTraverse(tile, frameState)) {
+        refines = await this.updateAndPushChildren(tile, frameState, stack);
+      }
+
+      // 3. decide if should render (select) this tile
+      //   - tile does not have render content
+      //   - tile has render content and tile is `add` type (pointcloud)
+      //   - tile has render content and tile is `replace` type (photogrammetry) and can't refine any further
+      const parent = tile.parent;
+      const parentRefines = !parent || parent._refines;
+      const stoppedRefining = !refines && parentRefines;
+
+      if (!tile.hasRenderContent) {
+        this.emptyTiles[tile.id] = tile;
+        this.loadTile(tile);
+        if (stoppedRefining) {
+          this.selectTile(tile, frameState);
+        }
+        // additive tiles
+      } else if (tile.refine === TILE3D_REFINEMENT.ADD) {
+        // Additive tiles are always loaded and selected
+        this.loadTile(tile, frameState);
+        this.selectTile(tile, frameState);
+
+        // replace tiles
+      } else if (tile.refine === TILE3D_REFINEMENT.REPLACE) {
+        // Always load tiles in the base traversal
+        // Select tiles that can't refine further
+        this.loadTile(tile, frameState);
+        if (stoppedRefining) {
+          this.selectTile(tile, frameState);
+        }
+      }
+
+      // 3. update cache, most recent touched tiles have higher priority to be fetched from server
+      this.touchTile(tile, frameState);
+
+      // 4. update tile refine prop
+      tile._refines = refines;
+    }
+  }
+
+  async updateChildTiles(tile, frameState) {
+    const children = tile.children;
+
+    for (const child of children) {
+      this.updateTile(child, frameState);
+    }
+  }
+
+  // eslint-disable-next-line complexity
+  async updateAndPushChildren(tile, frameState, stack) {
+    const {loadSiblings, skipLevelOfDetail} = this.options;
+
+    // update child tiles (visibility and expiration)
+    await this.updateChildTiles(tile, frameState);
+
+    const children = tile.children;
+
+    // sort children tiles
+    children.sort(this.compareDistanceToCamera);
+
+    // For traditional replacement refinement only refine if all children are loaded.
+    // Empty tiles are exempt since it looks better if children stream in as they are loaded to fill the empty space.
+    const checkRefines =
+      !skipLevelOfDetail && tile.refine === TILE3D_REFINEMENT.REPLACE && tile.hasRenderContent;
```

Comment:
Consider removing `skipLevelOfDetail` in this or part 2 since we would need to have loaders generate gpu commands so that it would look ok during loading.